File: go/task/07while.go
package task

import (
	"fmt"
	"math"

	"github.com/pkg/errors"
	"github.com/shahzodshafizod/ptaskbook/go/pkg/checker"
	"github.com/shahzodshafizod/ptaskbook/go/pkg/scanner"
)

type whileTask struct {
	dataPath string
	name     string
	count    int
	scanner  scanner.Scanner
	checker  checker.Checker
}

func NewWhileTask(pathPrefix string) Task {
	return &whileTask{
		dataPath: pathPrefix + "/07while/While%03d/%03d/%03d",
		name:     "While",
		count:    30,
	}
}

func (wt *whileTask) GetCount() int { return wt.count }

func (wt *whileTask) GetName() string { return wt.name }

func (wt *whileTask) SetData(taskNo int, testNo int) error {
	var filePath = fmt.Sprintf(wt.dataPath, taskNo, testNo, testNo)
	scanner, err := scanner.New(filePath + ".dat")
	if err != nil {
		return errors.Wrap(err, "scanner.New")
	}
	checker, err := checker.New(filePath + ".ans")
	if err != nil {
		return errors.Wrap(err, "checker.New")
	}
	wt.scanner = scanner
	wt.checker = checker
	return nil
}

func (wt *whileTask) ScannerEOF() bool { return wt.scanner.EOF() }

func (wt *whileTask) CheckerEOF() bool { return wt.checker.EOF() }

func (wt *whileTask) CleanData() {
	wt.scanner.RemoveFiles()
	wt.checker.RemoveFiles()
}

func (wt *whileTask) GetFn(taskNo int) func() bool {
	switch taskNo {
	case 1:
		return wt.while1
	case 2:
		return wt.while2
	case 3:
		return wt.while3
	case 4:
		return wt.while4
	case 5:
		return wt.while5
	case 6:
		return wt.while6
	case 7:
		return wt.while7
	case 8:
		return wt.while8
	case 9:
		return wt.while9
	case 10:
		return wt.while10
	case 11:
		return wt.while11
	case 12:
		return wt.while12
	case 13:
		return wt.while13
	case 14:
		return wt.while14
	case 15:
		return wt.while15
	case 16:
		return wt.while16
	case 17:
		return wt.while17
	case 18:
		return wt.while18
	case 19:
		return wt.while19
	case 20:
		return wt.while20
	case 21:
		return wt.while21
	case 22:
		return wt.while22
	case 23:
		return wt.while23
	case 24:
		return wt.while24
	case 25:
		return wt.while25
	case 26:
		return wt.while26
	case 27:
		return wt.while27
	case 28:
		return wt.while28
	case 29:
		return wt.while29
	case 30:
		return wt.while30
	default:
		return nil
	}
}

// { "no": 1, "dat": "2", "ans": "2" }
func (wt *whileTask) while1() bool {
	var a = wt.scanner.NextFloat64()
	var b = wt.scanner.NextFloat64()
	for a >= b {
		a -= b
	}
	var unusedPart = a
	return wt.checker.CompareFloat64(2, unusedPart)
}

// { "no": 2, "dat": "2", "ans": "" }
func (wt *whileTask) while2() bool {
	var a = wt.scanner.NextFloat64()
	var b = wt.scanner.NextFloat64()
	var segments = 0
	for a >= b {
		a -= b
		segments++
	}
	return wt.checker.CompareInt(segments)
}

// { "no": 3, "dat": "", "ans": "" }
func (wt *whileTask) while3() bool {
	var n = wt.scanner.NextInt()
	var k = wt.scanner.NextInt()
	var segments = 0
	for n >= k {
		n -= k
		segments++
	}
	var unusedPart = n
	return wt.checker.CompareInt(segments, unusedPart)
}

// { "no": 4, "dat": "", "ans": "" }
func (wt *whileTask) while4() bool {
	var n = wt.scanner.NextInt()
	var degree3 = 1
	for degree3 < n {
		degree3 *= 3
	}
	var isDegree3 = degree3 == n
	return wt.checker.CompareBool(isDegree3)
}

// { "no": 5, "dat": "", "ans": "" }
func (wt *whileTask) while5() bool {
	var n = wt.scanner.NextInt()
	var degree2 = 1
	var k = 0
	for degree2 < n {
		degree2 *= 2
		k++
	}
	return wt.checker.CompareInt(k)
}

// { "no": 6, "dat": "", "ans": "2" }
func (wt *whileTask) while6() bool {
	var n = wt.scanner.NextInt()
	var doubleFactorial = 1.0
	for n > 0 {
		doubleFactorial *= float64(n)
		n -= 2
	}
	return wt.checker.CompareFloat64(2, doubleFactorial)
}

// { "no": 7, "dat": "", "ans": "" }
func (wt *whileTask) while7() bool {
	var n = wt.scanner.NextInt()
	var k = 1
	for k*k <= n {
		k++
	}
	return wt.checker.CompareInt(k)
}

// { "no": 8, "dat": "", "ans": "" }
func (wt *whileTask) while8() bool {
	var n = wt.scanner.NextInt()
	var k = 1
	for k*k <= n {
		k++
	}
	k--
	return wt.checker.CompareInt(k)
}

// { "no": 9, "dat": "", "ans": "" }
func (wt *whileTask) while9() bool {
	var n = wt.scanner.NextInt()
	var k = 0
	var degree3 = 1
	for degree3 <= n {
		degree3 *= 3
		k++
	}
	return wt.checker.CompareInt(k)
}

// { "no": 10, "dat": "", "ans": "" }
func (wt *whileTask) while10() bool {
	var n = wt.scanner.NextInt()
	var k = 0
	var degree3 = 1
	for degree3 < n {
		degree3 *= 3
		k++
	}
	k--
	return wt.checker.CompareInt(k)
}

// { "no": 11, "dat": "", "ans": "" }
func (wt *whileTask) while11() bool {
	var n = wt.scanner.NextInt()
	var sum = 0
	var k = 0
	for sum < n {
		k++
		sum += k
	}
	return wt.checker.CompareInt(k, sum)
}

// { "no": 12, "dat": "", "ans": "" }
func (wt *whileTask) while12() bool {
	var n = wt.scanner.NextInt()
	var sum = 0
	var k = 0
	for sum <= n {
		k++
		sum += k
	}
	sum -= k
	k--
	return wt.checker.CompareInt(k, sum)
}

// { "no": 13, "dat": "5", "ans": "5" }
func (wt *whileTask) while13() bool {
	var a = wt.scanner.NextFloat64()
	var sum = 0.0
	var k = 0
	for sum <= a {
		k++
		sum += 1.0 / float64(k)
	}
	if !wt.checker.CompareInt(k) {
		return false
	}
	return wt.checker.CompareFloat64(5, sum)
}

// { "no": 14, "dat": "5", "ans": "5" }
func (wt *whileTask) while14() bool {
	var a = wt.scanner.NextFloat64()
	var sum = 0.0
	var k = 0
	for sum < a {
		k++
		sum += 1.0 / float64(k)
	}
	sum -= 1.0 / float64(k)
	k--
	if !wt.checker.CompareInt(k) {
		return false
	}
	return wt.checker.CompareFloat64(5, sum)
}

// { "no": 15, "dat": "2", "ans": "2" }
func (wt *whileTask) while15() bool {
	var amount = 1000.0
	var percent = wt.scanner.NextFloat64()
	var years = 0
	for amount <= 1100 {
		years++
		amount += amount * percent / 100
	}
	if !wt.checker.CompareInt(years) {
		return false
	}
	return wt.checker.CompareFloat64(2, amount)
}

// { "no": 16, "dat": "2", "ans": "3" }
func (wt *whileTask) while16() bool {
	var run = 10.0
	var percent = wt.scanner.NextFloat64()
	var days = 0
	var totalRun = 0.0
	for totalRun <= 200 {
		totalRun += run
		days++
		run += run * percent / 100
	}
	if !wt.checker.CompareInt(days) {
		return false
	}
	return wt.checker.CompareFloat64(3, totalRun)
}

// { "no": 17, "dat": "", "ans": "" }
func (wt *whileTask) while17() bool {
	var n = wt.scanner.NextInt()
	var digit int
	for n > 0 {
		digit = n % 10
		if !wt.checker.CompareInt(digit) {
			return false
		}
		n /= 10
	}
	return true
}

// { "no": 18, "dat": "", "ans": "" }
func (wt *whileTask) while18() bool {
	var n = wt.scanner.NextInt()
	var sum, count = 0, 0
	for n > 0 {
		count++
		sum += n % 10
		n /= 10
	}
	return wt.checker.CompareInt(count, sum)
}

// { "no": 19, "dat": "", "ans": "" }
func (wt *whileTask) while19() bool {
	var n = wt.scanner.NextInt()
	var chappa = 0
	for n > 0 {
		chappa = chappa*10 + n%10
		n /= 10
	}
	n = chappa
	return wt.checker.CompareInt(n)
}

// { "no": 20, "dat": "", "ans": "" }
func (wt *whileTask) while20() bool {
	var n = wt.scanner.NextInt()
	var hasTwo = false
	for n > 0 && !hasTwo {
		if n%10 == 2 {
			hasTwo = true
		}
		n /= 10
	}
	return wt.checker.CompareBool(hasTwo)
}

// { "no": 21, "dat": "", "ans": "" }
func (wt *whileTask) while21() bool {
	var n = wt.scanner.NextInt()
	var hasOdd = false
	for n > 0 && !hasOdd {
		if n%10%2 != 0 {
			hasOdd = true
		}
		n /= 10
	}
	return wt.checker.CompareBool(hasOdd)
}

// { "no": 22, "dat": "", "ans": "" }
func (wt *whileTask) while22() bool {
	var n = wt.scanner.NextInt()
	var divisor = int(math.Sqrt(float64(n)))
	var isPrime = true
	for divisor > 1 && isPrime {
		if n%divisor == 0 {
			isPrime = false
		}
		divisor--
	}
	return wt.checker.CompareBool(isPrime)
}

// { "no": 23, "dat": "", "ans": "" }
func (wt *whileTask) while23() bool {
	var a = wt.scanner.NextInt()
	var b = wt.scanner.NextInt()
	for b != 0 {
		a, b = b, a%b
	}
	return wt.checker.CompareInt(a)
}

// { "no": 24, "dat": "", "ans": "" }
func (wt *whileTask) while24() bool {
	var n = wt.scanner.NextInt()
	var prevF, nextF = 1, 1
	for nextF < n {
		prevF, nextF = nextF, prevF+nextF
	}
	var isFibonacci = nextF == n
	return wt.checker.CompareBool(isFibonacci)
}

// { "no": 25, "dat": "", "ans": "" }
func (wt *whileTask) while25() bool {
	var n = wt.scanner.NextInt()
	var prevF, nextF = 1, 1
	for nextF <= n {
		prevF, nextF = nextF, prevF+nextF
	}
	return wt.checker.CompareInt(nextF)
}

// { "no": 26, "dat": "", "ans": "" }
func (wt *whileTask) while26() bool {
	var n = wt.scanner.NextInt()
	var prevF, nextF = 1, 1
	for nextF < n {
		prevF, nextF = nextF, prevF+nextF
	}
	nextF += prevF
	return wt.checker.CompareInt(prevF, nextF)
}

// { "no": 27, "dat": "", "ans": "" }
func (wt *whileTask) while27() bool {
	var n = wt.scanner.NextInt()
	var prevF, nextF = 1, 1
	var k = 2
	for nextF < n {
		prevF, nextF = nextF, prevF+nextF
		k++
	}
	return wt.checker.CompareInt(k)
}

// { "no": 28, "dat": "6", "ans": "6" }
func (wt *whileTask) while28() bool {
	var eps = wt.scanner.NextFloat64()
	var prevA = 2.0
	var currA float64
	var k = 1
	for {
		currA = 2 + 1/prevA
		k++
		if math.Abs(currA-prevA) < eps {
			break
		}
		prevA = currA
	}
	if !wt.checker.CompareInt(k) {
		return false
	}
	return wt.checker.CompareFloat64(6, prevA, currA)
}

// { "no": 29, "dat": "6", "ans": "6" }
func (wt *whileTask) while29() bool {
	var eps = wt.scanner.NextFloat64()
	var prevA, currA = 1.0, 2.0
	var nextA float64
	var k = 1
	for {
		nextA = (prevA + 2*currA) / 3
		k++
		if math.Abs(prevA-currA) < eps {
			break
		}
		prevA, currA = currA, nextA
	}
	if !wt.checker.CompareInt(k) {
		return false
	}
	return wt.checker.CompareFloat64(6, prevA, currA)
}

// { "no": 30, "dat": "2", "ans": "" }
func (wt *whileTask) while30() bool {
	var a = wt.scanner.NextFloat64()
	var b = wt.scanner.NextFloat64()
	var c = wt.scanner.NextFloat64()
	var cInA = 0
	for a >= c {
		cInA++
		a -= c
	}
	var squaresCount = 0
	for b >= c {
		squaresCount += cInA
		b -= c
	}
	return wt.checker.CompareInt(squaresCount)
}

File: kt/task/07while.kt
package task

import kotlin.math.pow

class WhileTask {

    var datScanner = scanner.Scanner()
    var ansScanner = scanner.Scanner()

    fun solve(pathPrefix: String) {
        for (taskNo in 1..30) {
            var taskNoStr = "%03d".format(taskNo)
            var fileName: String
            for (testNo in 1..100) {
                var testNoStr = "%03d".format(testNo)
                fileName = "$pathPrefix/07while/While$taskNoStr/$testNoStr/$testNoStr"
                if (!datScanner.changeReader("$fileName.dat")) {
                    println("While$taskNoStr couldn't open datScanner source")
                    return
                }
                if (!ansScanner.changeReader("$fileName.ans")) {
                    println("While$taskNoStr couldn't open ansScanner source")
                    return
                }
                val ok = when (taskNo) {
                    1 -> this.while1()
                    2 -> this.while2()
                    3 -> this.while3()
                    4 -> this.while4()
                    5 -> this.while5()
                    6 -> this.while6()
                    7 -> this.while7()
                    8 -> this.while8()
                    9 -> this.while9()
                    10 -> this.while10()
                    11 -> this.while11()
                    12 -> this.while12()
                    13 -> this.while13()
                    14 -> this.while14()
                    15 -> this.while15()
                    16 -> this.while16()
                    17 -> this.while17()
                    18 -> this.while18()
                    19 -> this.while19()
                    20 -> this.while20()
                    21 -> this.while21()
                    22 -> this.while22()
                    23 -> this.while23()
                    24 -> this.while24()
                    25 -> this.while25()
                    26 -> this.while26()
                    27 -> this.while27()
                    28 -> this.while28()
                    29 -> this.while29()
                    30 -> this.while30()
                    else -> false
                }
                scanner.removeFiles()
                if (!ok) {
                    println("While$taskNoStr Test#$testNoStr has not been tested")
                    return
                }
                if (!datScanner.eof()) {
                    println("While$taskNoStr Test#$testNoStr input is not enough")
                    return
                }
                if (!ansScanner.eof()) {
                    println("While$taskNoStr Test#$testNoStr output is not enough")
                    return
                }
            }
            println("While$taskNoStr has been tested!")
        }
    }

    private fun while1(): Boolean {
        var a = this.datScanner.nextDouble()
        val b = this.datScanner.nextDouble()
        while (a >= b) a -= b
        return this.ansScanner.checkDouble(2, a)
    }

    private fun while2(): Boolean {
        var a = this.datScanner.nextDouble()
        val b = this.datScanner.nextDouble()
        var segmentsCount = 0
        while (a >= b) {
            a -= b
            segmentsCount++
        }
        return this.ansScanner.checkInt(segmentsCount)
    }

    private fun while3(): Boolean {
        var n = this.datScanner.nextInt()
        val k = this.datScanner.nextInt()
        var div = 0
        while (n >= k) {
            n -= k
            div++
        }
        val rem = n
        return this.ansScanner.checkInt(div, rem)
    }

    private fun while4(): Boolean {
        val n = this.datScanner.nextInt()
        var degree = 1
        while (degree < n) degree *= 3
        return this.ansScanner.checkBoolean(degree == n)
    }

    private fun while5(): Boolean {
        val n = this.datScanner.nextInt()
        var k = 0
        while (2.0.pow(k).toInt() < n) k++
        return this.ansScanner.checkInt(k)
    }

    private fun while6(): Boolean {
        val n = this.datScanner.nextInt()
        var doubleFact = 1.0
        for (index in n downTo 1 step 2) {
            doubleFact *= index
        }
        return this.ansScanner.checkDouble(2, doubleFact)
    }

    private fun while7(): Boolean {
        val n = this.datScanner.nextInt()
        var k = 1
        while (k*k <= n) {
            k++
        }
        return this.ansScanner.checkInt(k)
    }

    private fun while8(): Boolean {
        val n = this.datScanner.nextInt()
        var k = 0
        while ((k+1)*(k+1) <= n) {
            k++
        }
        return this.ansScanner.checkInt(k)
    }

    private fun while9(): Boolean {
        val n = this.datScanner.nextInt()
        var k = 0
        var degree3 = 1
        while (degree3 <= n) {
            degree3 *= 3
            k++
        }
        return this.ansScanner.checkInt(k)
    }

    private fun while10(): Boolean {
        val n = this.datScanner.nextInt()
        var k = -1
        var degree3 = 1
        do {
            degree3 *= 3
            k++
        } while (degree3 < n)
        return this.ansScanner.checkInt(k)
    }

    private fun while11(): Boolean {
        val n = this.datScanner.nextInt()
        var sum = 0
        var k = 0
        while (sum < n) {
            k++
            sum += k
        }
        return this.ansScanner.checkInt(k, sum)
    }

    private fun while12(): Boolean {
        val n = this.datScanner.nextInt()
        var sum = 0
        var k = 0
        while (sum + k + 1 <= n) {
            k++
            sum += k
        }
        return this.ansScanner.checkInt(k, sum)
    }

    private fun while13(): Boolean {
        val a = this.datScanner.nextDouble()
        var k = 0
        var sum = 0.0
        while (sum <= a) {
            k++
            sum += 1.0/k
        }
        return this.ansScanner.checkInt(k) && this.ansScanner.checkDouble(5, sum)
    }

    private fun while14(): Boolean {
        val a = this.datScanner.nextDouble()
        var k = 0
        var sum = 0.0
        while (sum < a) {
            k++
            sum += 1.0 / k
        }
        sum -= k
        k--
        return this.ansScanner.checkInt(k) && this.ansScanner.checkDouble(5, sum)
    }

    private fun while15(): Boolean {
        return false
    }

    private fun while16(): Boolean {
        return false
    }

    private fun while17(): Boolean {
        return false
    }

    private fun while18(): Boolean {
        return false
    }

    private fun while19(): Boolean {
        return false
    }

    private fun while20(): Boolean {
        return false
    }

    private fun while21(): Boolean {
        return false
    }

    private fun while22(): Boolean {
        return false
    }

    private fun while23(): Boolean {
        return false
    }

    private fun while24(): Boolean {
        return false
    }

    private fun while25(): Boolean {
        return false
    }

    private fun while26(): Boolean {
        return false
    }

    private fun while27(): Boolean {
        return false
    }

    private fun while28(): Boolean {
        return false
    }

    private fun while29(): Boolean {
        return false
    }

    private fun while30(): Boolean {
        return false
    }
}

/*
While1: 2
While6: 2
While13: 5
While14: 5
While15: 2
While16: 3
While28: 6
While29: 6
 */